Third party libraries such as psycopg2 marked as W7936(missing-manifest-dependency)

I tracked down an issue with a check returning erroneous result:

```
odoo/local-src/camptocamp_tools/utils.py:10: [W7936(missing-manifest-dependency), ] Missing external dependency "psycopg2" from manifest. More info: https://github.com/OCA/odoo-community.org/blob/master/website/Contribution/CONTRIBUTING.rst#external-dependencies
```

It happens here:

https://github.com/OCA/pylint-odoo/blob/67fa4893ce04b93e19612a577bc4563f0bfa62dc/pylint_odoo/checkers/modules_odoo.py#L356-L361

We add the list of modules to whitelist, psycopg2 is in the list, so we shouldn't have this warning. However, if you use `isort` to actually sort the imports, you may have a configuration file `.isort.cfg`. In this configuration, you will have the list of modules to consider as third party, so isort places them at the right place. Example:

```
[settings]
known_third_party = dateutil,openupgradelib,pkg_resources,psycopg2,requests,setuptools,urllib2,yaml
```

`isort.place_module()` builds a list of known patterns, and check first if the lib is a third party, returning this result first (if `psycopg2` is in the third party list and the std list, it will be considered as third party).

Refs (not in the most recent version, but easy to read): 
* https://github.com/timothycrosley/isort/blob/88a23e8aaacddf727ce01040dab31b9f7f55c497/isort/isort.py#L148-L155
* https://github.com/timothycrosley/isort/blob/88a23e8aaacddf727ce01040dab31b9f7f55c497/isort/isort.py#L257-L262

@guewen 
I got it
Thank for explanation.

I just reproduced using:
```sh
echo """[settings]
known_third_party = psycopg2
""" > ~/.isort.cfg

pylint --load-plugins=pylint_odoo -d all -e missing-manifest-dependency ./addons/point_of_sale
```

Output:
`************* Module point_of_sale.models.pos_order
./addons/point_of_sale/models/pos_order.py:7:0: W7936: Missing external dependency "psycopg2" from manifest. More info: https://github.com/OCA/odoo-community.org/blob/master/website/Contribution/CONTRIBUTING.rst#external-dependencies (missing-manifest-dependency)`

FYI I didn't find a easy fix since that the isort 4.3.4 is assigning the patterns to identify each library from the constructor `__init__` 😞:
 - https://github.com/timothycrosley/isort/blob/88a23e8aaacddf727ce01040dab31b9f7f55c497/isort/isort.py#L148-L155

And here is not easy to re-build them.

Newer versions 4.3.21 is using a separated class 😃 
 - https://github.com/timothycrosley/isort/blob/cd9c2f6e70196bd429d0179635e0fac2d034ac5e/isort/finders.py#L74-L86
Here it is good to re-build them.

But if we adds a patch just for newer versions locally will be raise an error for older ones.
Then it is more complicated than I expected to fix it.

--- a/pylint_odoo/checkers/modules_odoo.py
+++ b/pylint_odoo/checkers/modules_odoo.py
@@ -276,9 +276,18 @@ class ModuleChecker(misc.WrapperModuleChecker):
         node.file = self.linter.current_file
         self.inh_dup.setdefault(key, []).append(node)
 
+    def _build_whitelist_module_patterns(self):
+        known_patterns = []
+        for known_pattern in self.config.import_name_whitelist:
+            pattern = known_pattern.replace('*', '.*').replace('?', '.?')
+            known_patterns.append(re.compile('^' + pattern + '$'))
+        return known_patterns
+
     def open(self):
         """Define variables to use cache"""
         self.inh_dup = {}
+        patterns = self._build_whitelist_module_patterns()
+        self._whitelist_module_patterns = patterns
         super(ModuleChecker, self).open()
 
     def close(self):
@@ -339,6 +348,25 @@ class ModuleChecker(misc.WrapperModuleChecker):
         except:
             pass
 
+    def _is_module_name_in_whitelist(self, module_name):
+        # Try to find most specific placement instruction match (if any)
+        # (from isort place_module() method)
+        parts = module_name.split('.')
+        module_names_to_check = [
+            '.'.join(parts[:first_k])
+            for first_k in range(len(parts), 0, -1)
+        ]
+        # Check if one of the module name is part of the whitelist.
+        # For an module name such as 'anybox.testing.openerp', the
+        # modules names to check will be:
+        # ['anybox.testing.openerp', 'anybox.testing', 'anybox']
+        # Only one of them has to be in the whitelist to be accepted.
+        for module_name_to_check in module_names_to_check:
+            for pattern in self._whitelist_module_patterns:
+                if pattern.match(module_name_to_check):
+                    return True
+        return False
+
     def _check_imported_packages(self, node, module_name):
         """Check if the import node is a external dependency to validate it"""
         if not module_name:
@@ -353,10 +381,10 @@ class ModuleChecker(misc.WrapperModuleChecker):
         if self._is_absolute_import(node, module_name):
             # skip absolute imports
             return
-        isort_obj = isort.SortImports(
-            file_contents='',
-            known_standard_library=self.config.import_name_whitelist,
-        )
+        if self._is_module_name_in_whitelist(module_name):
+            # ignore whitelisted modules
+            return
+        isort_obj = isort.SortImports(file_contents='')
         import_category = isort_obj.place_module(module_name)
         if import_category not in ('FIRSTPARTY', 'THIRDPARTY'):
             # skip if is not a external library or is a white list library

